Integrate libvips

Rodrigodd opened the pull request:
This PR aims to integrate [libvips](https://github.com/libvips/libvips) into this project to get a better algorithm for downsampling images (mangas frequently use half-tone shading, which is very susceptible to aliasing when downsampling). As a bonus, it will allow moving all image format implementations upstream, decreasing the maintenance burden.

Roadmap:

- [x] Refactor CMake to use `ExternalProject` (as [Komelia](https://github.com/Snd-R/Komelia) currently does).
- [x] Write automated tests (kinda hard to know if all libraries are include in the final library without testing it directly).
  - [x] Run the tests on CI? I would need to setup a emulator there.
- [x] Link all libraries statically, if possible
- [ ] Integrate libvips (by copying Komelia, again). This will be done by adding a `decoder_vips` that accepts all formats it can, and implementing resize using a Lanczos filter or similar. This resize alone will not be enough to remove all aliasing; this will need to be addressed in real-time in subsampling-scale-image-view to fully fix it.
  - [x] Compile and link libvips
  - [x] Add `decoder_vips.h/cpp`
  - [x] Implement `cropBorders`
  - [x] Implement resize with filtering.
  - [x] Benchmark the implementation.
  - [ ] Handle ICC profiles properly (currently just converting to sRGB in RGBA8888 format).
- [x] Remove all no-longer-needed decoders, as libvips replaces them. Or let them remain, but disabled by default? Each decoder is currently guarded by an option, though I don't know what this is useful for.
  - [x] Restructure the library to be simpler and take more advantage of libvips.

Rodrigodd committed:
Refactor CMake project to use ExternalProject

Also rely on the existing build tool of each external project, instead
of patching CMake into them.

Most `ExternalProject_Add` wrappers were copied from Komelia, at
[https://github.com/Snd-R/Komelia/tree/8b938397e5c8743d4e995d2cc16203367de6f67d/image-decoder/native/cmake],
but modified to make sure all libraries are staticaly linked.

Rodrigodd committed:
ci: Inspect and archive library file

Also only run assembleDebug for now, full assemble is too slow.

Rodrigodd committed:
Add tests

Rodrigodd committed:
Add decoder libvips

Rodrigodd committed:
Install more recent version of meson

AntsyLich commented:
Leaving some comments:
1. We probably don't need to do a full copy of Komelia for the decoder methods and such but this can be discussed later.
2. For the last task I think we should drop the existing decoders in favor of vips

Rodrigodd committed:
Add documentation to decoder_base

Took me a while to figure out how the API work

Rodrigodd committed:
Remove all decoders minus libvips

AntsyLich commented:
Can you rename the packages from `tachiyomi.decoder` to `dev.mihon.image.decoder`

Rodrigodd commented:
@AntsyLich yes, I can. I will just do a final refactor of the code and submit a PR to your fork, with that change.

Rodrigodd committed:
Resize full image and then decode region

Was trying to optimize by cropping the image to the desired region, and
then resizing only that, but libvips seems to already be smart enough,
so that is not necessary.

Rodrigodd committed:
Write to Bitmap pixels directer

The intermediate buffer is not necessary, the `decode` function will
already take care of outputing in the correct format.

Rodrigodd committed:
Add documentation and minor refactors

Rodrigodd commented:
@AntsyLich should it really be `dev.mihon.image.decoder`? I think it would make more sense to be `app.mihon.image-decoder`, because you guys own [mihon.app](https://mihon.app/), and the project is already called `image-decoder`. Asking just to be sure, I am fine either way.

AntsyLich commented:
we also own [mihon.dev](https://mihon.dev) and packages can't have `-` in them. Alternatively we can use `_` but `image_decoder` doesn't sound pleasing or `imagedecoder`

Rodrigodd commented:
Ok, I will use `dev.mihon.image.decoder` then.

Rodrigodd committed:
Remove loops from tests

Was looping to detect memory corruption. Not needed anymore.

Rodrigodd committed:
Rename package to `dev.mihon.image.decoder`

And groupId/artifactId as `dev.mihon:image-decoder`

Rodrigodd committed:
Remove ImageInfo struct

Only `bounds` were being used, so I only keeped that.

Rodrigodd committed:
Rename `java/` to `kotlin/`

Rodrigodd committed:
Add .cache to gitignore

This folder is genrated by clangd

wwww-wwww committed:
lmcs2 for rgb, cmyk, no profile

Rodrigodd committed:
Add benchmark module

Rodrigodd committed:
Fix test path

Rodrigodd committed:
fix: Initialize VIPS_INIT only once

Mihon was crashing somethimes crashing when decoding the first image. I
notice Mihon was decoding two images at the same time, which made me
think it could be a threading issue. Reading libvips documentation, it
states that it is safe, but VIPS_INIT must only be called in a
single-thread. I was calling VIPS_INIT for every decoder created, which
I believe was the issue.

This commit moves VIPS_INIT to JNI_OnLoad, which is called only once
when the library is loaded.

Rodrigodd committed:
Update to Gradle 8.9 and NDK 26.3

Rodrigodd committed:
Refactor CMake project to use ExternalProject

Rodrigodd committed:
ci: Run tests on emulator

Rodrigodd committed:
ci: Remove build step before running emulator

The emulator is failing to create due to lack of space. Maybe the
previous build is using too much? I removed the build step, but it will
still be built after the emulator is created, let's see if it will still
break.

Rodrigodd committed:
ci: Free disk space

Rodrigodd committed:
ci: remove autoremove

Rodrigodd committed:
ci: Setup Android SDK

Rodrigodd committed:
ci: Run :benchmark in dryRunMode

This will just check if the benchmark can run without errors.

Rodrigodd committed:
Try to cache CMake output

I am not very confident that building with the cache will work, even
more if we modify the CMakeLists.txt file between builds. But waiting
>20min per build is very painful (and there are both Debug and Release
builds).

Rodrigodd committed:
Fix typo in ./gradlew -P arg

The typo came from Androids documentation...

Rodrigodd committed:
ci: Give up on testing bench on CI

Rodrigodd commented:
Superseded by https://github.com/mihonapp/mihon/pull/3388.
